subprojects/base-services/src/main/java/org/gradle/internal/service/DefaultServiceRegistry.java

```diff
@@ -1156,7 +1156,12 @@ private static void assertValidServiceType(Class<?> serviceClass) {
     private static String format(Type type) {
         if (type instanceof Class) {
             Class<?> aClass = (Class) type;
-            return aClass.getSimpleName();
+            Class<?> enclosingClass = aClass.getEnclosingClass();
+            if (enclosingClass != null) {
+                return format(enclosingClass) + "." + aClass.getSimpleName();
+            } else {
+                return aClass.getSimpleName();
+            }
```

Comment:
This makes the exceptions a bit more informative when you have enclosing classes named something simple like `Factory`.